local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,218 @@ fn test_slot_hash_expiry() {
+        make_identical_validator_configs(&default_config, node_stakes.len());
+    validator_configs[1].voting_disabled = true;
+
+    let mut config = ClusterConfig {
+        cluster_lamports: DEFAULT_CLUSTER_LAMPORTS + node_stakes.iter().sum::<u64>(),
+        node_stakes,
+        validator_configs,
+        validator_keys: Some(
+            validator_keys
+                .iter()
+                .map(|s| (s.clone(), true))
+                .collect::<Vec<_>>(),
+        ),
+        node_vote_keys: Some(node_vote_keys),
+        slots_per_epoch,
+        stakers_slot_offset: slots_per_epoch,
+        skip_warmup_slots: true,
+        ..ClusterConfig::default()
+    };
+    let mut cluster = LocalCluster::new(&mut config, SocketAddrSpace::Unspecified);
+
+    let common_ancestor_slot: u64 = num_a_leader_slots as u64 - 1;
+
+    let a_ledger_path = cluster.ledger_path(&a_pubkey);
+    let b_ledger_path = cluster.ledger_path(&b_pubkey);
+
+    // Immediately kill B and C (we just needed it for the initial stake distribution)
+    info!("Killing B and C");
+    let mut b_info = cluster.exit_node(&b_pubkey);
+    cluster.exit_node(&c_pubkey);
+
+    // Let A run for a while until we get to the common ancestor
+    info!("Run A on majority fork until it creates its own fork");
+    let now = Instant::now();
+    let mut last_vote_on_a;
+    loop {
+        last_vote_on_a = wait_for_last_vote_in_tower_to_land_in_ledger(&a_ledger_path, &a_pubkey);
+        if last_vote_on_a >= common_ancestor_slot {
+            let blockstore = open_blockstore(&a_ledger_path);
+            info!(
+                "A majority fork: {:?} common_ancestor_slot: {}",
+                AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>(),
+                common_ancestor_slot,
+            );
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+        sleep(Duration::from_millis(100));
+    }
+
+    // Copy A's ledger to B but truncate any blocks past the common ancestor.
+    {
+        info!("Copying A's ledger to B");
+        let blockstore_a = open_blockstore(&a_ledger_path);
+        let blockstore_b = open_blockstore(&b_ledger_path);
+        copy_blocks(common_ancestor_slot, &blockstore_a, &blockstore_b);
+    }
+
+    // Kill A and restart B with voting. B should now fork off
+    info!("Killing A");
+    let a_info = cluster.exit_node(&a_pubkey);
+
+    info!("Restarting B");
+    b_info.config.voting_disabled = false;
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    // Allow B to fork now.
+    info!("Allowing B to fork");
+    let now = Instant::now();
+    loop {
+        sleep(Duration::from_millis(200));
+        let (last_vote, _) = last_vote_in_tower(&b_ledger_path, &b_pubkey).unwrap();
+        if last_vote > (2 * num_a_leader_slots + num_b_leader_slots) as u64 {
```

Comment:
why is this `2 * num_a_leader_slots + num_b_leader_slots`, is this just to give the validator `B` enough time to vote on the last block in its 4 block set? Seems more readable check that `last_vote % (num_a_leader_slots + num_b_leader_slots) == 0`

Reply:
Because last_vote_on_a is normally on the second round of leader_schedule. This check is to make sure B is voting on its own leader block. But we don't need to wait for end of its leader block.